File: residential-units.R
# Make a single table of all residential units in the district.

# Load our spreadsheet of use codes.
usecodes = read.csv("use-codes.csv")

# FUNCTIONS

is_residential_rental_use_code <- function(usecode) {
	# Is the use code in our use-codes.csv file of residential use codes?
	# (use codes appear as "11" or "011", so always turn
	# into an integer first)
	return(sum(as.numeric(usecodes$usecode)==as.numeric(usecode)) > 0);
}

get_use_code_unit_count <- function(usecode) {
	# How many units can we assume are in a structure with this use code?
	units <- usecodes[usecodes$usecode==usecode,]$units;
	if (length(units) == 0) {
		print("invalid call to get_use_code_unit_count")
		print(use_code);
		return(NA);
	} else if (units == "") {
		# don't know
		return(NA);
	} else if (units == "1") {
		return(1);
	} else if (units == "2-4") {
		return(3);
	} else if (units == "5") {
		return(5);
	} else {
		print("invalid units data in use-codes.csv")
		print(use_code);
		print(units)
		return(NA);
	}
}

over_rows <- function(dataframe, func) {
	# Call func on each row of dataframe.
	#print(names(dataframe))
	for (i in 1:nrow(dataframe))
		func(dataframe[i,]);
}

append_unit <- function(cama_row, i, units, source_file) {
	# Make a row in residential_units from a row in
	# the source CAMA file, plus information about
	# which "unit" this is (i) when we automatically
	# expand a structure into its units.
	newrow = list(
		# property metadata
		square_suffix_lot=as.character(cama_row$SSL),
		use_code=cama_row$USECODE,

		# fields related to rent stabilization policy
		year_built=cama_row$AYB,
		owner_name=as.character(cama_row$OWNERNAME),
		structure_units=units,

		# address, unit, and geographic location
		address=as.character(cama_row$PREMISEADD),
		unit_number=expanded_unit_number(cama_row$UNITNUMBER, units, i),
		longitude=cama_row$X,
		latitude=cama_row$Y,

		# source information for this record
		source_file=source_file,
		source_object_id=cama_row$OBJECTID
	)

	# rbind() is extremely slow. To make this faster, we should pre-allocate
	# space in the data frame and then fill it in.
	if (residential_units_count == nrow(residential_units)) {
		# Allocate more space by doubling the size of the data frame.
		# We'll chop it down to the number of rows we actually filled
		# in at the end.
		space_to_add = (1+residential_units_count) # residential_units_count is zero in the first iteration, so at 1 so we allocate something
		residential_units <<- rbind(residential_units, data.frame(
			square_suffix_lot=character(space_to_add),
			use_code=numeric(space_to_add),
			year_built=numeric(space_to_add),
			owner_name=character(space_to_add),
			structure_units=numeric(space_to_add),
			address=character(space_to_add),
			unit_number=character(space_to_add),
			longitude=numeric(space_to_add),
			latitude=numeric(space_to_add),
			source_file=character(space_to_add),
			source_object_id=numeric(space_to_add),
			stringsAsFactors=F
		))
		print(residential_units_count)
	}

	# Copy into the data frame and increment the counter of the number
	# of rows we've actually used.
	residential_units[residential_units_count,] <<- as.data.frame(newrow, stringsAsFactors=F);
	residential_units_count <<- residential_units_count + 1; # "<<-" assigns to a global variable
	#print(residential_units)
};

expanded_unit_number <- function(source_unit, nunits, i) {
	if (nunits == 1) {
		return(source_unit);
	} else if (source_unit == "") {
		return(i);
	} else {
		return(paste(source_unit, "/", i))
	}
}

# MAIN

# Create an empty data frame to hold all of the residential units in DC.
residential_units_count = 0;
residential_units <- data.frame(
	square_suffix_lot=character(),
	use_code=numeric(),

	year_built=numeric(),
	owner_name=character(),
	structure_units=numeric(),

	address=character(),
	unit_number=character(),
	longitude=numeric(),
	latitude=numeric(),

	source_file=character(),
	source_object_id=numeric(),
		
	stringsAsFactors=F);

# Loop through the residential properties. Each property
# may have more than one unit, which we expand out.
over_rows(read.csv("data/cama_residential.csv"), function(row) {
	# Skip if not a residential / plausibly rental building.
	if (!is_residential_rental_use_code(row$USECODE)) { return(); }

	# There is no unit count in the dataset, so we have to guess
	# from the use code.
	units = get_use_code_unit_count(row$USECODE);
	if (is.na(units)) {
		print("no unit count available");
		print(row);
		return();
	}
	
	# Expand this building into units.
	for (i in 1:units)
		append_unit(row, i, units, "cama_residential");
})

# Loop through the condo properties. Each property
# may have more than one unit, which we expand out.
over_rows(read.csv("data/cama_condominium.csv"), function(row) {
	# Skip if not a residential / plausibly rental building.
	if (!is_residential_rental_use_code(row$USECODE)) { return(); }

	# There is no unit count in the dataset, so we have to guess
	# from the use code.
	units = get_use_code_unit_count(row$USECODE);
	if (is.na(units)) {
		print("no unit count available");
		print(row);
		return();
	}
	
	# Expand this building into units.
	for (i in 1:units)
		append_unit(row, i, units, "cama_condominium");
})

# Loop through the commercial properties. Each property
# may have more than one unit, which we expand out.
over_rows(read.csv("data/cama_commercial.csv"), function(row) {
	# Skip if not a residential / plausibly rental building.
	if (!is_residential_rental_use_code(row$USECODE)) { return(); }

	if (!is.na(row$NUM_UNITS)) {
		# If given, use the NUM_UNITS column.
		units = row$NUM_UNITS;
		if (units == 0) {
			print("zero units?");
			print(row);
			return();
		}
	} else {
		units = get_use_code_unit_count(row$USECODE);
		if (is.na(units)) {
			print("no unit count available");
			print(row);
			return();
		}
	}
	
	# Expand this building into units.
	for (i in 1:units)
		append_unit(row, i, units, "cama_commercial");
})

# Save!
residential_units = residential_units[0:residential_units_count,]
write.table(residential_units, "data/all_residential_units.csv", row.names=F)

File: residential-units.py
import csv
import random
import shapefile # pip3 install pyshp

try:
	# tqdm provides a nice progress meter
	from tqdm import tqdm
except ImportError:
	def tqdm(_, *args, **kwargs): return _


# Our information about use codes. Map use codes (as integers) to
# the CSV row about it.
use_codes = csv.DictReader(open("use-codes.csv"))
use_codes = { int(rec["usecode"]): rec for rec in use_codes }

# Owner strings.
government_owners = ("DISTRICT OF COLUMBIA HOUSING AUTHORITY", "UNITED STATES OF AMERICA", "DISTRICT OF COLUMBIA")
corporate_owner_suffixes = ("LLC","LP","INC","TRUSTEE","ASSOCIATES","PARTNERSHIP","CORPORATION","TRUSTEES","ASSOCIATION","TRUST","COMPANY","PRTNSHP","LTD","VENTURE","CORP","L.P.","LTP","UNIVERSITY","PROPERTY","PTSP","CORPORTATION","PROPERTIES","LLP","HOUSING","APARTMENTS","PART","FOUNDATION","CHURCH")

# Pre-load the ward and SMD boundaries.
geometries = { }
for ty, fn in (("ward", "Ward__2012"), ("smd", "Single_Member_District__2013")):
	sf = shapefile.Reader("data/%s.shp" % fn)
	for i, shape in enumerate(sf.iterShapes()):
		id = sf.record(i)[1]
		assert shape.shapeType == 5
		geometries.setdefault(ty, {})[id] = {
			"bbox": shape.bbox,
			"points": shape.points,
		}

# Output file.
residential_units = csv.writer(open("data/all-residential-units.csv", "w"))
residential_units.writerow([
	"square_suffix_lot", "use_code",
	"year_built", "owner_name", "structure_units", "unit_in_structure",
	"address", "unit_number", "longitude", "latitude", "ward", "smd",
	"source_file", "source_object_id",
	"govowned", "corpowned", "cooperative",
])

# Some functions.
def is_positive_number(num):
	try:
		return int(num) > 0
	except ValueError:
		# not an integer-looking string
		return False

def make_unit_num(original_unit_num, fake_unit_num, fake_unit_count):
	if fake_unit_count == 1:
		return original_unit_num
	elif original_unit_num.strip() == "":
		return fake_unit_num
	else:
		return original_unit_num + "/" + str(fake_unit_num)

def peek_bom(f):
	# The files from data.dc.gov have a byte order marker.
	# Read it so the CSV reader doesn't see it.
	f.read(1)
	return f

def find_containing_geometry(x, y, geometries):
	# Find which polygon in geometries contains the lnglat point.
	# This is not an efficient way to do this --- we should use
	# something with a spatial index. But we don't have so much
	# data that it's worth optimizing at all. We can just check
	# each polygon to see a) if its bounding box contains the point
	# and if so 2) if the poylgon contains the point.
	matches = set()
	for id, geom in geometries.items():
		if (geom["bbox"][0] <= x <= geom["bbox"][2]) and (geom["bbox"][1] <= y <= geom["bbox"][3]):
			if point_in_poly(x, y, geom["points"]):
				matches.add(id)
	assert len(matches) == 1 # point should be in exactly one polygon
	return list(matches)[0]

def point_in_poly(x,y,poly):
	# http://geospatialpython.com/2011/01/point-in-polygon.html
    n = len(poly)
    inside = False
    p1x,p1y = poly[0]
    for i in range(n+1):
        p2x,p2y = poly[i % n]
        if y > min(p1y,p2y):
            if y <= max(p1y,p2y):
                if x <= max(p1x,p2x):
                    if p1y != p2y:
                        xints = (y-p1y)*(p2x-p1x)/(p2y-p1y)+p1x
                    if p1x == p2x or x <= xints:
                        inside = not inside
        p1x,p1y = p2x,p2y
    return inside

# Process each file.
for fn in ("cama_residential.csv", "cama_condominium.csv", "cama_commercial.csv"):
	for row in tqdm(list(csv.DictReader(peek_bom(open("data/" + fn)))), desc=fn):
		# Is this record for a residential use?
		if not is_positive_number(row["USECODE"]) or int(row["USECODE"]) not in use_codes:
			continue

		# How many units do we think 
		use_code_info = use_codes[int(row["USECODE"])]
		if fn == "cama_commercial.csv" and is_positive_number(row["NUM_UNITS"]):
			# For the commercial file, use the NUM_UNITS column
			# if it has sane data.
			unit_count = int(row["NUM_UNITS"])
			exploded = True
		elif use_code_info["units"] == "1":
			unit_count = 1
			exploded = (fn != "cama_condominium.csv") # don't treat condo units as "exploded" from a structure
		elif use_code_info["units"] == "2-4":
			# We don't know how many units this structure has but it's between 2 and 4.
			# We could take a best guess, like 3, but lumping everything into a single
			# value makes further analysis look funky. So we'll spread it out between
			# 2 and 4. Based on some post-hoc analysis, it looks like 2 and 3 units are
			# more common than 4 -- it makes the results smoother when comparing to
			# 5.
			unit_count = random.choice([2,2,2,2,2,2,3,3,3,4])
			exploded = True
		elif use_code_info["units"] == "5":
			unit_count = 5
			exploded = True
		else:
			print("dont know how many units in", row)
			continue

		# Find the ward and SMD of this unit/building.
		row["X"] = float(row["X"])
		row["Y"] = float(row["Y"])
		ward = find_containing_geometry(row["X"], row["Y"], geometries["ward"])
		smd = find_containing_geometry(row["X"], row["Y"], geometries["smd"])

		# Write a record for each presumed unit in the record.
		for i in range(unit_count):
			residential_units.writerow([
				row["SSL"], int(row["USECODE"]),
				row["AYB"], row["OWNERNAME"], (unit_count if exploded else "NA"), ((i+1) if exploded else "NA"),
				row["PREMISEADD"], row["UNITNUMBER"],
				row["X"], row["Y"],
				ward, smd,
				fn, row["OBJECTID"],

				row["OWNERNAME"] in government_owners, # govowned
				row["OWNERNAME"].split(" ")[-1] in corporate_owner_suffixes, # last word of owner appears in corporate_owner_suffixes
				int(row["USECODE"]) in (26,27,126,127), # cooperative based on use code
			])
